graspy/inference/nonpar.py

```diff
@@ -0,0 +1,137 @@
+import numpy as np
+from scipy.spatial.distance import pdist
+
+from .base import BaseInference
+from ..embed import AdjacencySpectralEmbed, LaplacianSpectralEmbed
+from ..embed import OmnibusEmbed, select_dimension
+from ..utils import import_graph, is_symmetric, symmetrize
+
+
+class NonparametricTest(BaseInference):
+    """
+    Two sample hypothesis test for the nonparamatric problem of determining
+    whether two random dot product graphs have the same latent positions.
+
+    Parameters
+    ----------
+    embedding : string, { 'ase' (default), 'lse', 'omnibus'}
+        String describing the embedding method to use.
+        Must be one of:
+        'ase'
+            Embed each graph separately using adjacency spectral embedding
+            and use Procrustes to align the embeddings.
+        'lse'
+            Embed each graph separately using laplacian spectral embedding
+            and use Procrustes to align the embeddings.
+        'omnibus'
+            Embed all graphs simultaneously using omnibus embedding.
+
+    n_components : None (default), or Int
+        Number of embedding dimensions. If None, the optimal embedding
+        dimensions are found by the Zhu and Godsi algorithm.
+
+    n_bootstraps : 200 (default), or Int
+        Number of bootstrap iterations.
+
+    monte_iter : 1000 (default), or Int
+        Number of monte carlo iterations.
+    """
+
+    def __init__(self,
+                 embedding='ase',
+                 n_components=None,
+                 n_bootstraps=200,
+                 n_monte=1000):
+        if type(n_bootstraps) is not int:
+            raise TypeError()
+
+        if n_bootstraps < 1:
+            msg = '{} is invalid number of bootstraps, must be greater than 1'
+            raise ValueError(msg.format(n_bootstraps))
+
+        super().__init__(
+            embedding=embedding,
+            n_components=n_components,
+        )
+        # self.embedding = embedding
+        # self.n_components = n_components
+        self.n_bootstraps = n_bootstraps
+        self.n_monte = n_monte
+
+    def _bootstrap(self, x, y):
+        z = np.concatenate((x, y))
+        boots = []
+        for _ in range(self.n_bootstraps):
+            np.random.shuffle(z)
+            u = self._u_from_z(z, len(x))
+            boots.append(u)
+        return boots
+
+    def _u_from_z(self, z, n):
+        k = self._dist_matrix(z)
+        gk = np.exp(-k / np.median(k))
+        u = np.mean(gk[:n, :n]) + np.mean(gk[n:, n:]) - 2 * np.mean(gk[:n, n:])
+        return u
+
+    def _dist_matrix(self, z):
+        dists = pdist(z, 'euclidean')
+        zlen = len(z)
+        ind = np.triu_indices(zlen, k=1)
+        k = np.zeros((zlen, zlen))
+        k[ind] = dists
+        return symmetrize(k)
+
+    def _embed(self, A1, A2):
```

Comment:
If this is supposed to be the exact same as semipar we should refactor to the base class. Note that I don't think LSE makes sense for semipar so I took it out, have you tested it for nonpar?

Reply:
You're absolutely right